### PR TITLE
SORQA-39 Enter an exposure data in Case Directory

### PR DESCRIPTION
**Pull request by mk-sgent**
SORQA-39

SORQA-39DE

Jenkins:
**First**
https://jenkins.sormas.netzlink.com/job/sormas-RunE2eTests-branch/475/allure/#suites/ff98ba16245987a8dbc94846ba9dce63/51896d8ec64d1b10/
**Second**
https://jenkins.sormas.netzlink.com/job/sormas-RunE2eTests-branch/477/
**Thirth**
https://jenkins.sormas.netzlink.com/job/sormas-RunE2eTests-branch/502/allure/

image

### CODE REVIEW

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/ContinentUUIDs.java [hunk 1]
```diff
@@ -33,4 +35,8 @@
   public String toString() {
     return this.option;
   }
+
+  public static String getRandomContinentUUIDsName() {
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/ContinentUUIDs.java:39]: please remove, there is only one continent for now so we don't need this

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/CountryUUIDs.java [hunk 2]
```diff
@@ -33,4 +35,8 @@
   public String toString() {
     return this.option;
   }
+
+  public static String getRandomCountryUUIDsName() {
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/CountryUUIDs.java:39]: same as above

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/DistrictsValues.java [hunk 3]
```diff
@@ -49,4 +51,8 @@ public static String getNameFor(String option) {
     }
     throw new Exception("Unable to find " + option + " value in District Enum");
   }
+
+  public static String getRandomDistrictsValuesName() {
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/DistrictsValues.java:55]: same

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/RegionsValues.java [hunk 4]
```diff
@@ -26,12 +27,17 @@
...
   private final String name;
   private final String uuid;
+  private static Random random = new Random();
 
   RegionsValues(String name, String uuid) {
     this.name = name;
     this.uuid = uuid;
   }
 
+  public static String getRandomRegionsValuesName() {
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/RegionsValues.java:37]: same

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/SubcontinentUUIDs.java [hunk 5]
```diff
@@ -18,19 +18,27 @@
...
+  SubcontinentUUIDs(String name, String uuid) {
+    this.name = name;
+    this.uuid = uuid;
   }
 
-  @Override
-  public String toString() {
-    return this.option;
+  @SneakyThrows
+  public static String getCaptionFor(String option) {
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/SubcontinentUUIDs.java:37]: this needs to be refactored everywhere

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/TypeOfActivity.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package org.sormas.e2etests.enums;
+
+public enum TypeOfActivity {
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/enums/TypeOfActivity.java:3]: we already have:
**public enum ActivityTypes**
Please rename new created to reflect the need, and update also the existing one

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/cases/CaseDirectoryPage.java [hunk 7]
```diff
@@ -125,5 +125,7 @@
   public static final By INVESTIGATION_DISCARDED_BUTTON = By.id("Investigation discarded");
   public static final By DATE_FROM_COMBOBOX = By.cssSelector("#dateFrom input");
   public static final By DATE_TO_COMBOBOX = By.cssSelector("#dateTo input");
+  public static final By CASES_EPIDEMIOLOGICAL_DATA = By.cssSelector("#tab-cases-epidata");
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/cases/CaseDirectoryPage.java:128]: rename it accordingly

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/cases/CaseDirectoryPage.java [hunk 8]
```diff
@@ -125,5 +125,7 @@
...
+  public static final By CASES_NEW_ENTRY_EPIDEMIOLOGICAL_DATA = By.id("actionNewEntry");
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/cases/CaseDirectoryPage.java:129]: rename it accordingly

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/contacts/ContactDirectoryPage.java [hunk 9]
```diff
@@ -48,6 +48,31 @@
   public static final By GRID_HEADERS = By.xpath("//thead//tr//th");
   public static final String RESULTS_GRID_HEADER = "//div[contains(text(), '%s')]";
   public static final By FIRST_CONTACT_ID_BUTTON = By.cssSelector(".v-grid-row-has-data a[title]");
+  public static final By CONTACT_EPIDEMIOLOGICAL_DATA = By.cssSelector("#tab-contacts-epidata");
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/contacts/ContactDirectoryPage.java:51]: rename it accordingly

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/contacts/ContactDirectoryPage.java [hunk 10]
```diff
@@ -48,6 +48,31 @@
...
+  public static final By CASE_LIKE = By.id("contactOrCaseLike");
+  public static final By NEW_ENTRY_EPIDEMIOLOGICAL_DATA = By.id("actionNewEntry");
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/contacts/ContactDirectoryPage.java:53]: rename it accordingly

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/contacts/ContactDirectoryPage.java [hunk 11]
```diff
@@ -48,6 +48,31 @@
   public static final By GRID_HEADERS = By.xpath("//thead//tr//th");
   public static final String RESULTS_GRID_HEADER = "//div[contains(text(), '%s')]";
   public static final By FIRST_CONTACT_ID_BUTTON = By.cssSelector(".v-grid-row-has-data a[title]");
+  public static final By CONTACT_EPIDEMIOLOGICAL_DATA = By.cssSelector("#tab-contacts-epidata");
+  public static final By CASE_LIKE = By.id("contactOrCaseLike");
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/contacts/ContactDirectoryPage.java:52]: rename it accordingly

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/contacts/ContactDirectoryPage.java [hunk 12]
```diff
@@ -48,6 +48,31 @@
...
+      By.cssSelector(".v-window input#facilityDetails");
+  public static final By CONTACT_PERSON_FIRST_NAME =
+      By.cssSelector(".v-window input#contactPersonFirstName");
+  public static final By CONTACT_PERSON_LAST_NAME =
+      By.cssSelector(".v-window input#contactPersonLastName");
+  public static final By CONTACT_PERSON_PHONE_NUMBER =
+      By.cssSelector(".v-window input#contactPersonPhone");
+  public static final By CONTACT_PERSON_EMAIL_ADRESS =
+      By.cssSelector(".v-window input#contactPersonEmail");
   public static final By EPIDEMIOLOGICAL_DATA_TAB = By.cssSelector("#tab-contacts-epidata");
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/pages/application/contacts/ContactDirectoryPage.java:76]: please review all locators and rename them to refelct the WebElement type. So the ending should reflect the item in UI (eg: button, input, tab etc)

File: sormas-e2e-tests/src/main/java/org/sormas/e2etests/services/ContactService.java [hunk 13]
```diff
@@ -151,9 +159,78 @@ public Exposure buildGeneratedExposureDataForContact() {
...
+        .facility("Other facility")
+        .facilityDetails(faker.chuckNorris().fact())
+        .contactPersonFirstName(firstName)
+        .contactPersonLastName(lastName)
+        .contactPersonPhone(faker.phoneNumber().phoneNumber())
+        .contactPersonEmail(firstName + lastName + emailDomain)
+        .build();
+  }
+
+  public Exposure buildGeneratedExposureDataContactForRandomInputs() {
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/main/java/org/sormas/e2etests/services/ContactService.java:189]: what do you mean random inputs? what's the need of this method?

File: sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/cases/CaseDirectorySteps.java [hunk 14]
```diff
@@ -48,6 +110,19 @@ public CaseDirectorySteps(
       AssertHelpers assertHelpers,
       Faker faker) {
 
+    When(
+        "I select Type of activity except Gathering and Other options from Combobox in Exposure form",
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/cases/CaseDirectorySteps.java:114]: if we perform one action, why do we involve in the bdd sentence other items?

File: sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/contacts/ContactDirectorySteps.java [hunk 15]
```diff
@@ -208,32 +294,157 @@ public ContactDirectorySteps(
...
+          selectAllTypeOfPlace();
+        });
+
+    When(
+        "I fill Location form for Type of place by options excluded Other and Facility",
+        () -> {
+          webDriverHelpers.selectFromCombobox(TYPE_OF_PLACE_COMBOBOX, TypeOfPlace.HOME.getPlace());
+          exposureData = contactService.buildGeneratedExposureDataContactForRandomInputs();
+          fillLocation(exposureData);
+          webDriverHelpers.clickOnWebElementBySelector(DONE_BUTTON);
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/contacts/ContactDirectorySteps.java:380]: move out this action so we can reuse it after

File: sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/contacts/ContactDirectorySteps.java [hunk 16]
```diff
@@ -208,32 +294,157 @@ public ContactDirectorySteps(
...
+    When(
+        "I fill Location form for Type of place by Other option",
+        () -> {
+          webDriverHelpers.selectFromCombobox(TYPE_OF_PLACE_COMBOBOX, TypeOfPlace.OTHER.getPlace());
+          exposureData = contactService.buildGeneratedExposureDataContactForRandomInputs();
+          fillLocation(exposureData);
+          webDriverHelpers.fillInWebElement(
+              TYPE_OF_PLACE_DETAILS, exposureData.getTypeOfPlaceDetails());
+          webDriverHelpers.waitForPageLoaded();
+          webDriverHelpers.clickOnWebElementBySelector(DONE_BUTTON);
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/contacts/ContactDirectorySteps.java:392]: same

File: sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/contacts/ContactDirectorySteps.java [hunk 17]
```diff
@@ -208,32 +294,157 @@ public ContactDirectorySteps(
...
+        "I fill Location form for Type of place by options excluded Other and Facility",
+        () -> {
+          webDriverHelpers.selectFromCombobox(TYPE_OF_PLACE_COMBOBOX, TypeOfPlace.HOME.getPlace());
+          exposureData = contactService.buildGeneratedExposureDataContactForRandomInputs();
+          fillLocation(exposureData);
+          webDriverHelpers.clickOnWebElementBySelector(DONE_BUTTON);
+        });
+
+    When(
+        "I fill Location form for Type of place by Other option",
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/contacts/ContactDirectorySteps.java:384]: this method actually generates some random data and fills 1 field.
I don't see the need of building a new DTO only to update one field value.

File: sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/contacts/ContactDirectorySteps.java [hunk 18]
```diff
@@ -208,32 +294,157 @@ public ContactDirectorySteps(
...
+
+    When(
+        "I select a Type of activity ([^\"]*) option in Exposure form",
+        (String option) -> {
+          webDriverHelpers.waitForPageLoaded();
+          webDriverHelpers.selectFromCombobox(TYPE_OF_ACTIVITY_COMBOBOX, option);
+        });
+
+    When(
+        "I fill a Type of activity details in Exposure by ([^\"]*) TEXT",
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/contacts/ContactDirectorySteps.java:437]: do we need the TEXT keyword?

File: sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/contacts/ContactDirectorySteps.java [hunk 19]
```diff
@@ -208,32 +294,157 @@ public ContactDirectorySteps(
...
+        (String typeOfActivityDetails) -> {
+          webDriverHelpers.waitForPageLoaded();
+          webDriverHelpers.fillInWebElement(TYPE_OF_ACTIVITY_DETAILS, typeOfActivityDetails);
+        });
+
+    When(
+        "I fill a type of gathering details in Exposure form by ([^\"]*) TEXT",
```
Comment (rdutu-vg) [inline sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/contacts/ContactDirectorySteps.java:444]: same

**Review by rdutu-vg — CHANGES_REQUESTED** [review]

**Comment by rdutu-vg** [comment]
There are changes which can cause failures into multiple tests.
Please address them, then aside test passing proof, attach also a regression run.

File: sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/cases/EpidemiologicalDataCaseSteps.java [hunk 20]
```diff
@@ -96,6 +120,58 @@ public EpidemiologicalDataCaseSteps(
...
+    //              SOURCE_CASE_WINDOW_FIRST_RESULT_OPTION);
+    //          webDriverHelpers.clickOnWebElementBySelector(
+    //              SOURCE_CASE_CONTACT_WINDOW_FIRST_RESULT_OPTION);
+    //
+    // webDriverHelpers.waitForRowToBeSelected(SOURCE_CASE_CONTACT_WINDOW_FIRST_RESULT_OPTION);
+    //          webDriverHelpers.waitUntilElementIsVisibleAndClickable(
+    //              SOURCE_CASE_CONTACT_WINDOW_CONFIRM_BUTTON);
+    //
+    // webDriverHelpers.clickOnWebElementBySelector(SOURCE_CASE_CONTACT_WINDOW_CONFIRM_BUTTON);
+    //        });
```
Comment (jp-sgent) [inline sormas-e2e-tests/src/test/java/org/sormas/e2etests/steps/web/application/cases/EpidemiologicalDataCaseSteps.java:174]: I don't think we should be committing commented-out code.

File: sormas-e2e-tests/src/test/resources/features/sanity/web/EpidemiologicalDataCase.feature [hunk 21]
```diff
@@ -57,6 +57,110 @@ Feature: Epidemiological data coverage
...
+    Then I click on New Entry in Exposure Details Known in Cases directory
+    And I tick a Probable infection environmental box in Exposure for Epidemiological data tab in Cases
+    And I select from Combobox all options in Type of activity field in Exposure for Epidemiological data tab for Cases for DE version
+    Then I select a Type of activity Sonstiges option in Exposure for Epidemiological data tab in Cases
+    And I fill a Type of activity details field in Exposure for Epidemiological data tab in Cases
+    Then I select a Type of activity Versammlung option in Exposure for Epidemiological data tab in Cases
+    And I select from Combobox all Type of gathering in Exposure for Epidemiological data tab in Cases for DE version
+    And I select a type of gathering Sonstiges option from Combobox in Exposure for Epidemiological data tab in Cases
+    And I fill a type of gathering details in Exposure for Epidemiological data tab in Cases
+    And I fill Location form for Type of place field by "Unbekannt" options in Case directory for DE version
```
Comment (jp-sgent) [inline sormas-e2e-tests/src/test/resources/features/sanity/web/EpidemiologicalDataCase.feature:131]: We were instructed not to use a different language in step definitions - even as a parameter.
This applies to all german words below as well.

**Comment by rdutu-vg** [comment]
@mk-sgent please run also a regression over this branch to check how stable are the changes

**Comment by mk-sgent** [comment]
https://jenkins.sormas.netzlink.com/job/sormas-RunE2eTests-branch/501/allure/

JobsForSorqa-39

**Review by rdutu-vg — APPROVED** [review]

**Pull request merged**